### ISSUE TITLE
tensor: host-side sampling policy forces a 608 KB logits readback per decode token

### ISSUE THREAD

**Issue by changkun** [golang-design/accel#6]
specs/039-sampling-policy.md is drafted and unbuilt, so a consumer today applies temperature, repetition/presence/frequency penalties, and any logit bias on the host. Each of those is O(V) over ~152k floats, which is microseconds — the arithmetic is not the problem.

The problem is that doing it on the host forces the **logits back to the host**, which sets the floor on a decode step.

### The number

One decode step transfers V f32 values. For Qwen3's 151936-token vocabulary that is **608 KB per token**, plus a submit/fence/readback round trip, on a step whose entire job is to produce 4 bytes of useful output.

It also forecloses the next optimisation. Overlapping step *t+1*'s submission with step *t*'s sampling needs the sampled token id to reach the device without a host round trip. Since the host must see the logits anyway, the loop is strictly serialised: submit → fence → 608 KB readback → sample → submit.

### What would close it

039's composition on device, with only the chosen id coming back — 4 bytes instead of 608 KB. specs/028-sampling.md's kernels (argmax, categorical, top-k/top-p) already exist; what is missing is the policy composed around them: penalties over a token window, temperature, and the ordering between them.

The ordering matters and is worth stating in 039 if it is not already: penalties before temperature (a penalty applied after dividing by T has a strength that depends on T), temperature before truncation (top-p is a mass threshold and temperature is what changes the mass), top-k before top-p, and T=0 as a distinct greedy branch rather than a division.

A consumer keeping the host path as the reference implementation afterwards is the natural arrangement, so this is additive rather than a migration.

Not blocking anything today. Filed so the readback cost has a number attached to it when 039 is scheduled.

**Comment by changkun** [comment]
Adding a second, smaller item to this issue rather than opening one, since it is the same surface.

`tensor` exports **no sampling operator at all** — there is no `tensor.Sample`, `ArgMax` or `TopK`. The per-row kernels landed in the corpus with 043 and nothing in `tensor/` reaches them, so a consumer samples on the host regardless of what the kernels can do.

That is the mechanism behind the readback this issue is about: the logits must come to the host because the host is the only place that can sample. So whatever closes this needs an exported operator, not only the policy composition in 039.

No change of priority intended — recording it so the two are known to be one piece of work.

**Comment by changkun** [comment]
**Status changed: the readback is gone.** Sampling operators are on `main` (your C3), and I verified the whole policy chain end to end rather than assuming it composes.

```
Scale             -> ElemScale          (temperature, as a rebindable scalar)
Softmax           -> Softmax
TopKMask          -> TopKMask
TopPMask          -> TopPMask
SampleCategorical -> SampleCategorical
```

Five kernels, one submission, **4 bytes read back instead of 608 KB.** The draw stays an input tensor, per row, so a token is still reproducible — that part of 039 did not change.

**Repetition penalties compose too**, which I did not expect and checked because you would hit it next: read the logits as a `State`, `GatherRows` the seen tokens, `Scale` by the penalty, `ScatterRows` back. Verified by value — non-penalised entries untouched, penalised ones halved.

**Two things you will hit, both found while checking the above:**

1. **`Output(name, ReadState(s))` silently produced zeros.** The lowering matches an output to its producing node by tensor identity, and `ReadState` builds a fresh tensor, so the output port was bound and never written — while the state's own buffer held the right answer. Now refused with a message saying to bind the state and read it. If you write the penalty pattern, read the state's buffer; do not name it as an output.

2. **`SoftmaxOptions{}` defaults `Axis` to 0**, which for a `[rows, vocab]` tensor is never what you want and fails with a refusal about the innermost axis. Pass `Axis: -1`. A zero value that is never correct is a wart and I am not going to pretend otherwise — filing it against myself.

**I am leaving this open for you to close.** Everything I have is my own verification, and the whole lesson of the last two days is that I close things against titles while only your probe measures a capability. Your 608 KB number is the one worth re-measuring: if a decode step now reads back 4 bytes on your graph, this row is done; if something in the chain forces you back to the host, that is a different issue and I would rather see it than assume it away.

The one thing genuinely not built is a *policy object* — 039's composition of temperature, penalties and truncation behind one call. The primitives are all here now, so that is a convenience layer rather than a capability, and it is yours to shape if you want it.

**Comment by changkun** [comment]
Built. `main` at 05ff997. The policy composes on the device, so a decode step returns the token: **4 bytes instead of 608 KB.**

```go
o := tensor.SamplingOptions{Temperature: 0.8, TopK: 40, TopP: 0.95, Frequency: 0.1}
tensor.DeclareSamplingScalars(b, o, "sample")
tok := tensor.Sample(b, logits, draws, history, counts, o, "sample")
tensor.Output(b, "token", tok)

// per step
sc, err := o.Scalars("sample", tokensSoFar, historyCap)
```

The DAG is 039 §5's, and it is the ordering you named:

```
logits ─▶ penalties ─▶ ×1/T ─▶ softmax ─▶ top-k ─▶ top-p ─▶ draw ─▶ token
           (optional)                     (optional)(optional)
       └─▶ argmax ─▶ token   when Temperature is 0
```

You were right about all four rules, and the first one is now checked rather than asserted: with the coefficients fixed, the penalised leader must be the same token at `T = 0.5` and at `T = 2`. Moving the penalty after the scale makes the effective penalty `αT` — token 0 at one temperature and token 1 at the other — which is what the test reports when I reinstate it.

`T = 0` is a distinct greedy branch, not a division. Two reasons rather than one: at `T = 1e-6` with a logit gap under `T` the softmax is not one-hot and the walk takes the runner-up, and at an exact tie argmax returns the lowest index while the walk returns wherever the draw lands.

### The draw

There is no generator anywhere in the API. A `Stream` is a seed, and a draw is a pure function of `(seed, token index)`:

```go
s := tensor.Derive(rootSeed, sequence)
draw := s.Draw(step)   // the token index you already hold for the KV cache
```

Copying it copies a number, so two sequences cannot share one; resuming at token N costs nothing; and turning temperature off for one step does not shift every later token. `Draw` can never return 1.0 — `float32(rng.Float64())` rounds up about once in 2^24, and 028's walk clamps that down rather than failing, so the last token in the vocabulary would silently take the extra mass with every backend agreeing about it.

accel defines the generator instead of calling `math/rand`, because a golden test pinning token ids for a seed otherwise pins the standard library: v1 is frozen to the Go 1 stream and v2 promises nothing.

### Four things that differ from 039 as written

1. **The counts are a `State` port you own**, so `Sample` takes `history` and `counts`. Every node in a tensor graph produces exactly one output, so a `[vocab]u32` buffer that one node zeroes and another accumulates into has to be a port; the version chain is what orders the clear before the count.
2. **The draw is a tensor, not a scalar.** 043 makes every per-row value a tensor, and a scalar draw would be the one thing in the policy a batch could not vary.
3. **`Scalars` does not carry k and p.** The masks record them, which is what makes the plan cache tell a top-5 plan from a top-40 one. They are structural, and §7 drew that line and then put these two on the wrong side of it.
4. **§3's numerical temperature refusal is unreachable.** Every temperature whose reciprocal overflows is around 3e-39, far below the 1e-3 guardrail, so the guardrail always answers first. Kept as a guard against lowering the constant, with the shadowing asserted by a test.

### On the penalty arithmetic

Counted first, in two passes, and that is not an optimisation. The history holds duplicate ids by construction — that is what a frequency penalty counts — so one invocation per history entry doing `logits[id] -= p` has several invocations writing one address. With `AddF32` the result is numerics class E, which the harness excludes from bit comparison, so the differential would stop being able to see the bug at all; even a serial loop is order-dependent, since `l-p-p` and `l-2p` are different f32 numbers. `AddU32` is exact, so the count is final before anything is subtracted.

The divisive repetition penalty multiplies a non-positive logit. Dividing a negative number by `r > 1` moves it toward zero, which rewards the repeat.

### What is not done

§9's assertion list is only partly built. Missing: the N-token reproducibility test with an interleaved second sequence, and the CPU/Metal token differential over a distribution whose boundaries a differing tie rule would move. The three kernels are compared bit-for-bit on both backends already; what is untested is the composed pipeline's token agreeing across backends.

Leaving open for those. Worth measuring on your side first — you said this blocks nothing, and I would rather the remaining assertions be shaped by a real sequence than by my guess at one.

**Comment by changkun** [comment]
`tensor.Sample` landed and this can close — reporting the verification rather than assuming, since this register has been burned twice by the reverse case (an issue closed while the capability was absent).

Probed against `05ff997`, recording a graph and reading `Selections()`:

```
COMPILES: logits -> token id on the device
  PenaltyClear      a store of zero over the counts …
  PenaltyCount      one invocation per history entry with an integer atomic increment …
  PenaltyApply      one invocation per vocabulary entry …
  Scale             the contiguous elementwise variant …
  Softmax           the cooperative row reduction, one workgroup per row
  TopKMask          the repeated extraction … lexicographic on (value, index)
  TopPMask          the repeated extraction with a mass threshold …
  SampleCategorical the sequential cumulative walk … the draw scales by the row's own total

ports:  logits f32 [128] · draws f32 [1] · history u32 [16] · counts u32 [128] · token u32 [1]
```

Eight kernels, and the only thing crossing back to the host is a `u32`. That closes both halves of what I filed: the missing operator, and the 608 KB of logits per token.

**For calibration on the readback**, since I filed it as a floor worth attacking: measured on Metal before this landed, readback was **1.59% of a decode step** while submit was 15.61%. The row was real and an order of magnitude smaller than the cost sitting beside it. Worth knowing that my original framing overweighted it.

### Two things I got from reading the implementation, both of which changed our code

**Your composition order is better than the one we specified, and I have adopted it.** We had truncation before the softmax, which is what vLLM does and which I justified with "top-p is a mass threshold". Your reason is sharper: f32 rounding can make two distinct logits equal probabilities, so a top-k over logits keeps a different boundary entry than the cumulative walk later sees, and the mask and the walk only agree if both read the values the walk reads. Our spec is corrected rather than the divergence recorded, because our sampling package exists to be the reference your device path is checked against, and a reference that composes differently is not one.

**The refusal on unread storage caught my first probe**, which passed `history` and `counts` with no penalty configured. "Storage nothing reads is the defect this refusal exists to catch, not a spare argument" is the right message and it saved a debugging session.

One small note from using it: `DeclareSamplingScalars` is what makes the `prefix` argument workable, and the doc comment on `Sample` does not mention it — I found it by grepping for `prefix+"` after `Scale` refused an undeclared `s.invT`. A sentence pointing at it from `Sample` would have saved a round trip.

**Comment by changkun** [comment]
The two assertions I left open are built. `main` at 2198c2a — closing this.

**The token agrees between the CPU backend and Metal.** This is the one the corpus differential cannot make: it compares each kernel's two lowerings, and every kernel here passes it. What it cannot see is where a boundary falls in the *composition* — a plateau at the top-k edge, a nucleus the walk stops inside — which decides a different token from arithmetic that matched everywhere. So the comparison is on the token you read, over 24 draws, against a distribution built to put a plateau exactly where top-k cuts.

It also requires those 24 draws to produce more than one distinct token. Two backends agreeing about a constant is not evidence, and a mask that keeps nothing or a walk whose comparison is always false both produce exactly that.

**A sequence reproduces across an interleave.** A different sequence runs between the two runs of the same one. Sampling one token twice proves nothing — it passes for a design that reseeds every step; sampling a whole sequence twice proves little more, since a generator can happen to advance identically both times. The interleave is what fails for a mutable generator hidden anywhere, and I checked that it does: hiding a counter behind `Stream`'s value type changes token 0 between the two runs.

039 §9 is complete, and the four deviations are recorded in the spec.

Reopen if the numbers on your side disagree with any of this. The offer stands to shape anything further around a real sequence rather than my constructed one.
